security/advancedtls/crl.go

```diff
@@ -0,0 +1,489 @@
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package advancedtls
+
+import (
+	"bytes"
+	"crypto/sha1"
+	"crypto/tls"
+	"crypto/x509"
+	"crypto/x509/pkix"
+	"encoding/asn1"
+	"encoding/binary"
+	"encoding/hex"
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"path/filepath"
+	"strings"
+	"time"
+
+	"log"
+)
+
+// Cache is an interface to cache CRL files.
+// All caches from golang-lru satisfy the interface.
+type Cache interface {
+	Add(key, value interface{}) bool
+	Contains(key interface{}) bool
+	Get(key interface{}) (value interface{}, ok bool)
+}
+
+// Config contains options for CRL lookup
+type Config struct {
+	// RootDir is the directory to search for CRL files.
+	// Directory format must match OpenSSL X509_LOOKUP_hash_dir(3).
+	RootDir string
+	// AllowUndetermined controls if certificate chains with UNDETERMINED
+	// revocation status are allowed to complete.
+	AllowUndetermined bool
+	// Cache will store CRL files if not nil, otherwise files are reloaded for every lookup.
+	Cache Cache
+}
+
+// RevocationStatus is the revocation status for a certificate or chain
+type RevocationStatus int
+
+const (
+	// UNDETERMINED means we couldn't find or verify a CRL for the cert
+	UNDETERMINED RevocationStatus = iota
+	// UNREVOKED means we found the CRL for the cert and the cert is not revoked
+	UNREVOKED
+	// REVOKED means we found the CRL and the cert is revoked
+	REVOKED
+)
+
+func (s RevocationStatus) String() string {
+	return [...]string{"UNDETERMINED", "UNREVOKED", "REVOKED"}[s]
+}
+
+// certificateListExt contains a pkix.CertificateList and parsed
+// extensions that aren't provided by the golang CRL parser
+type certificateListExt struct {
+	CertList *pkix.CertificateList
+	// RFC5280, 5.2.1, all conforming CRLs must have a AKID with the ID method
+	AuthorityKeyID []byte
+}
+
+const tagDirectoryName = 4
+
+var (
+	// RFC5280, 5.2.4 id-ce-deltaCRLIndicator OBJECT IDENTIFIER ::= { id-ce 27 }
+	oidDeltaCRLIndicator = asn1.ObjectIdentifier{2, 5, 29, 27}
+	// RFC5280, 5.2.5 id-ce-issuingDistributionPoint OBJECT IDENTIFIER ::= { id-ce 28 }
+	oidIssuingDistributionPoint = asn1.ObjectIdentifier{2, 5, 29, 28}
+	// RFC5280, 5.3.3 id-ce-certificateIssuer   OBJECT IDENTIFIER ::= { id-ce 29 }
+	oidCertificateIssuer = asn1.ObjectIdentifier{2, 5, 29, 29}
+	// RFC5290, 4.2.1.1 id-ce-authorityKeyIdentifier OBJECT IDENTIFIER ::=  { id-ce 35 }
+	oidAuthorityKeyIdentifier = asn1.ObjectIdentifier{2, 5, 29, 35}
+)
+
+// X509NameHash implements the OpenSSL X509_NAME_hash function for hashed directory lookups.
+func X509NameHash(r pkix.RDNSequence) string {
```

Comment:
Is this function supposed to be public? I am wondering if I need to use it while integrating to advanced-tls. Thanks!

Reply:
I don't think you would need it, I can make it private